misk/src/main/kotlin/misk/web/jetty/WebActionsServlet.kt

```diff
@@ -107,33 +114,88 @@ internal class WebActionsServlet @Inject constructor(
   }
 }
 
+private fun callback(response: ServletUpgradeResponse): UpstreamResponse {
+  return object : UpstreamResponse {
+    override var statusCode: Int
+      get() = response.statusCode
+      set(value) {
+        response.statusCode = value
+      }
+
+    override val headers: Headers
+      get() = response.headers()
+
+    override fun setHeader(name: String, value: String) {
+      response.setHeader(name, value)
+    }
+
+    override fun addHeaders(headers: Headers) {
+      for (i in 0 until headers.size()) {
+        response.addHeader(headers.name(i), headers.value(i))
+      }
+    }
+  }
+}
+
+private fun callback(response: HttpServletResponse): UpstreamResponse {
+  return object : UpstreamResponse {
+    override var statusCode: Int
+      get() = response.status
+      set(value) {
+        response.status = value
+      }
+
+    override val headers: Headers
+      get() = response.headers()
+
+    override fun setHeader(name: String, value: String) {
+      response.setHeader(name, value)
+    }
+
+    override fun addHeaders(headers: Headers) {
+      for (i in 0 until headers.size()) {
+        response.addHeader(headers.name(i), headers.value(i))
+      }
+    }
+  }
+}
+
 internal fun HttpServletRequest.contentType() = contentType?.let { MediaType.parse(it) }
 
+internal fun ServletUpgradeResponse.headers(): Headers {
+  val result = Headers.Builder()
+  for (name in headerNames) {
+    for (value in getHeaders(name)) {
+      result.add(name, value)
+    }
+  }
+  return result.build()
+}
+
 internal fun HttpServletRequest.headers(): Headers {
-  val headersBuilder = Headers.Builder()
-  val headerNames = headerNames
-  for (headerName in headerNames) {
-    val headerValues = getHeaders(headerName)
-    for (headerValue in headerValues) {
-      headersBuilder.add(headerName, headerValue)
+  val result = Headers.Builder()
+  for (name in headerNames) {
+    for (value in getHeaders(name)) {
+      result.add(name, value)
     }
   }
-  return headersBuilder.build()
+  return result.build()
 }
 
-private fun HttpServletRequest.urlString(): HttpUrl? {
-  return if (queryString == null)
-    HttpUrl.parse(requestURL.toString()) else
-    HttpUrl.parse(requestURL.toString() + "?" + queryString)
+internal fun HttpServletResponse.headers(): Headers {
+  val result = Headers.Builder()
+  for (name in headerNames) {
+    for (value in getHeaders(name)) {
+      result.add(name, value)
+    }
+  }
+  return result.build()
 }
 
-private fun HttpServletRequest.asRequest(): Request {
-  return Request(
-      urlString()!!,
-      dispatchMechanism(),
-      headers(),
-      inputStream.source().buffer()
-  )
+internal fun HttpServletRequest.httpUrl(): HttpUrl {
+  return if (queryString == null)
+    HttpUrl.get("$requestURL") else
+    HttpUrl.get("$requestURL?$queryString")
```

Comment:
I'm not use to this kind of indention for if/else statements..

Reply:
Fixed.